In [1]:
from os import chdir 
chdir('./lib')

In [2]:
!pwd

/Users/ginodefalco/dsi/dsi_repo/DSI_SM_3/projects/project-05/lib


In [3]:
from project_5 import general_process, load_data_from_database, make_data_dict, general_model, general_transformer

# Step 1 - Benchmarking

**NOTE: EACH OF THESE SHOULD BE WRITTEN SOLELY WITH REGARD TO STEP 1 - BENCHMARKING**

### Domain and Data


MADELON is an artificial dataset, which was part of the NIPS 2003 feature selection challenge. This is a two-class classification problem with continuous input variables. The difficulty is that the problem is multivariate and highly non-linear.
MADELON is an artificial dataset containing data points grouped in 32 clusters placed on the vertices of a five dimensional hypercube and randomly labeled +1 or -1. The five dimensions constitute 5 informative features. 15 linear combinations of those features were added to form a set of 20 (redundant) informative features. Based on those 20 features one must separate the examples into the 2 classes (corresponding to the +-1 labels). We added a number of distractor feature called 'probes' having no predictive power. The order of the features and patterns were randomized.

### Problem Statement

We need to demonstrate a capacity to identify relevant features of this data set using machine learning. However, in this first step, we are establishing a benchmark.

### Solution Statement


In order to establish a benchmark for our problem, we need to connect and access the MADELON database and query the relevant data to return from the dataframe. From this dataframe we will make a data dictionary by splitting the data into a feature matrix and target matrix using train_test_split. After that, we will write a function that will transform the data, in this case we will use Standard Scaler. And, finally, we will write another function that will run a general regression model on the transformed data from the previous step.

### Metric

Our metric is our accuracy score from our training and test data.

### Benchmark

Our benchmark was our accuracy score from our general logistic regression with a high C value(1000) in order to perform minimal regularization: 78.8% train score and 51.6% test score.

## Implementation

Implement the following code pipeline using the functions you write in `lib/project_5.py`.

<img src="assets/benchmarking.png" width="600px">

In [4]:
# Let's use the function we wrote to load our data into the dataframe
df = load_data_from_database()
df.shape

(2000, 502)

In [5]:
# Let's get info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Columns: 502 entries, index to label
dtypes: int64(502)
memory usage: 7.7 MB


In [6]:
# Describe the Data
df.describe()

index     feat_000     feat_001     feat_002     feat_003  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    999.500000   481.722500   483.452500   510.166000   483.384500   
std     577.494589     6.421769    30.186294    38.899165     9.059895   
min       0.000000   462.000000   381.000000   370.000000   453.000000   
25%     499.750000   477.000000   464.000000   485.000000   477.000000   
50%     999.500000   482.000000   483.000000   510.500000   483.000000   
75%    1499.250000   486.000000   503.000000   536.000000   490.000000   
max    1999.000000   503.000000   600.000000   654.000000   519.000000   

          feat_004     feat_005     feat_006     feat_007     feat_008  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    501.612500   479.259000   480.109500   476.565000   486.793500   
std      41.389418     6.795956    40.575925     1.384461    15.043836   
min     371.000000   459.000000   334.000000   471.000000   430.000000   
25%     475.000000   475.000000   452.750000   476.000000   477.000000   
50%     500.000000   479.000000   480.000000   477.000000   487.000000   
75%     528.000000   484.000000   506.250000   477.000000   496.250000   
max     688.000000   505.000000   611.000000   481.000000   536.000000   

          ...         feat_491     feat_492     feat_493    feat_494  \
count     ...      2000.000000  2000.000000  2000.000000  2000.00000   
mean      ...       478.811500   486.356500   496.565500   493.49950   
std       ...         4.011735    23.967366   127.635442    34.81902   
min       ...       463.000000   391.000000   130.000000   368.00000   
25%       ...       476.000000   471.000000   404.000000   470.00000   
50%       ...       479.000000   486.000000   504.000000   492.00000   
75%       ...       481.000000   502.000000   586.000000   517.00000   
max       ...       497.000000   566.000000   920.000000   615.00000   

          feat_495     feat_496     feat_497     feat_498     feat_499  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    510.893000   478.219500   483.309000   507.977000   490.266000   
std      37.459353     5.880613    13.559847    37.224297    25.825273   
min     398.000000   457.000000   435.000000   363.000000   403.000000   
25%     486.000000   474.000000   474.000000   482.000000   473.000000   
50%     511.000000   478.000000   483.000000   508.000000   490.000000   
75%     535.000000   482.000000   492.000000   533.000000   507.250000   
max     661.000000   500.000000   535.000000   644.000000   583.000000   

            label  
count  2000.00000  
mean      0.00000  
std       1.00025  
min      -1.00000  
25%      -1.00000  
50%       0.00000  
75%       1.00000  
max       1.00000  

[8 rows x 502 columns]

In [7]:
# Let's load our make_data_dict function and run a train_test_split on the data
df = make_data_dict(df, random_state=0)
df

{'X_test':       index  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
 349     349       480       488       498       481       481       476   
 433     433       474       491       506       489       503       483   
 1863   1863       473       482       438       484       511       478   
 157     157       475       516       531       476       490       496   
 1295   1295       489       481       511       485       472       475   
 220     220       488       484       497       466       529       474   
 499     499       474       503       572       472       491       483   
 961     961       497       493       537       505       542       479   
 856     856       477       503       547       492       483       468   
 644     644       494       503       526       471       444       475   
 870     870       485       464       436       474       502       478   
 1025   1025       483       489       516       479       486       490   
 9

In [8]:
# Let's standardize our data
from sklearn.preprocessing import StandardScaler
df = general_transformer(StandardScaler(), df)
df

{'X_test': array([[-1.13953285, -0.27435433,  0.11906386, ..., -1.1556245 ,
         -1.84664861,  0.73678577],
        [-0.99429601, -1.20087927,  0.21810034, ..., -0.4053489 ,
          1.31542093,  2.14088061],
        [ 1.47818828, -1.3553001 , -0.0790091 , ..., -1.00556938,
          0.44312588, -0.31628536],
        ..., 
        [-0.96317383, -0.42877515, -0.0790091 , ...,  1.39531255,
         -0.10205852,  0.03473835],
        [ 0.88340884,  0.65217061, -0.64021582, ..., -0.48037646,
         -0.23835462, -0.94032752],
        [-1.5804304 , -0.11993351,  1.70364753, ..., -0.25529378,
          0.57942198, -0.00426429]]),
 'X_train': array([[ 0.48054951, -0.7376168 ,  0.51520977, ..., -1.75584499,
         -0.72902059, -1.25234859],
        [ 0.31629356, -0.58319598,  0.84533137, ...,  0.3449267 ,
         -0.64724292,  2.72592013],
        [-0.5240053 , -0.11993351,  0.54822193, ..., -1.08059694,
         -1.05613123,  0.11274362],
        ..., 
        [-0.53610837,  1.733116

In [9]:
# We are running a simple logistic regression with a large C value to  perform minimum regularization
# and set a benchmark score
from sklearn.linear_model import LogisticRegression
df = general_model(LogisticRegression(C=100000), df)

In [10]:
df

{'X_test': array([[-1.13953285, -0.27435433,  0.11906386, ..., -1.1556245 ,
         -1.84664861,  0.73678577],
        [-0.99429601, -1.20087927,  0.21810034, ..., -0.4053489 ,
          1.31542093,  2.14088061],
        [ 1.47818828, -1.3553001 , -0.0790091 , ..., -1.00556938,
          0.44312588, -0.31628536],
        ..., 
        [-0.96317383, -0.42877515, -0.0790091 , ...,  1.39531255,
         -0.10205852,  0.03473835],
        [ 0.88340884,  0.65217061, -0.64021582, ..., -0.48037646,
         -0.23835462, -0.94032752],
        [-1.5804304 , -0.11993351,  1.70364753, ..., -0.25529378,
          0.57942198, -0.00426429]]),
 'X_train': array([[ 0.48054951, -0.7376168 ,  0.51520977, ..., -1.75584499,
         -0.72902059, -1.25234859],
        [ 0.31629356, -0.58319598,  0.84533137, ...,  0.3449267 ,
         -0.64724292,  2.72592013],
        [-0.5240053 , -0.11993351,  0.54822193, ..., -1.08059694,
         -1.05613123,  0.11274362],
        ..., 
        [-0.53610837,  1.733116

In [11]:
df['train_score'], df['test_score']

(0.77200000000000002, 0.502)

# Step 1 Conclusion:

We built pipeline to perform a naive logistic regression as a baseline model
and set a high C value in order to perform minimal regularization. Our benchmark score was:
77.2% on the train
50.2% on the test
 - Looks like we overfit a bit